### ISSUE TITLE
Rule Change: no-shadow - exception for HOF

### ISSUE THREAD

**Issue by ximex** [eslint/eslint#16575]
### What rule do you want to change?

no-shadow

### What change to do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
// should fail
const items = [1, 2, 3, 4];
const item = 1;
const _item = items.filter((item) => (item % 2));

// should pass
const items = [1, 2, 3, 4];
const item = items.filter((item) => (item % 2));
```

### What does the rule currently do for this code?

I think no-shadow shouldn't show an error/warning if the variable name of the result of a higher order function (HOF) is the same as a local variable in the HOF as it is declared after the processing if the HOF. At least there should be an option to disable it.

### What will the rule do after it's changed?

It won't show a warning/error if a variable name is used in a HOF and the variable the results gets assigned to

### Participation

- [ ] I am willing to submit a pull request to implement this change.

### Additional comments

https://github.com/typescript-eslint/typescript-eslint/issues/6069

**Comment by yeonjuan** [comment]
Hi @ximex. There is an `ignoreOnInitialization` option. (see [no-shadow](https://eslint.org/docs/latest/rules/no-shadow#ignoreoninitialization)). Maybe, Is this the option you want ?

> The ignoreOnInitialization option is false by default. If it is true, it prevents reporting shadowing of variables in their initializers when the shadowed variable is presumably still uninitialized.
The shadowed variable must be on the left side. The shadowing variable must be on the right side and declared in a callback function or in an IIFE.

- [demo](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXNoYWRvdzogW1wiZXJyb3JcIiwge1wiaWdub3JlT25Jbml0aWFsaXphdGlvblwiOiB0cnVlfV0gKi9cblxuLy8gc2hvdWxkIGZhaWxcbmNvbnN0IGl0ZW1zMSA9IFsxLCAyLCAzLCA0XTtcbmNvbnN0IGl0ZW0xID0gMTtcbmNvbnN0IF9pdGVtMSA9IGl0ZW1zMS5maWx0ZXIoKGl0ZW0xKSA9PiAoaXRlbTEgJSAyKSk7XG5cbi8vIHNob3VsZCBwYXNzXG5jb25zdCBpdGVtczIgPSBbMSwgMiwgMywgNF07XG5jb25zdCBpdGVtMiA9IGl0ZW1zMi5maWx0ZXIoKGl0ZW0yKSA9PiAoaXRlbTIgJSAyKSk7Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJjb25zdHJ1Y3Rvci1zdXBlciI6WyJlcnJvciJdLCJmb3ItZGlyZWN0aW9uIjpbImVycm9yIl0sImdldHRlci1yZXR1cm4iOlsiZXJyb3IiXSwibm8tYXN5bmMtcHJvbWlzZS1leGVjdXRvciI6WyJlcnJvciJdLCJuby1jYXNlLWRlY2xhcmF0aW9ucyI6WyJlcnJvciJdLCJuby1jbGFzcy1hc3NpZ24iOlsiZXJyb3IiXSwibm8tY29tcGFyZS1uZWctemVybyI6WyJlcnJvciJdLCJuby1jb25kLWFzc2lnbiI6WyJlcnJvciJdLCJuby1jb25zdC1hc3NpZ24iOlsiZXJyb3IiXSwibm8tY29uc3RhbnQtY29uZGl0aW9uIjpbImVycm9yIl0sIm5vLWNvbnRyb2wtcmVnZXgiOlsiZXJyb3IiXSwibm8tZGVidWdnZXIiOlsiZXJyb3IiXSwibm8tZGVsZXRlLXZhciI6WyJlcnJvciJdLCJuby1kdXBlLWFyZ3MiOlsiZXJyb3IiXSwibm8tZHVwZS1jbGFzcy1tZW1iZXJzIjpbImVycm9yIl0sIm5vLWR1cGUtZWxzZS1pZiI6WyJlcnJvciJdLCJuby1kdXBlLWtleXMiOlsiZXJyb3IiXSwibm8tZHVwbGljYXRlLWNhc2UiOlsiZXJyb3IiXSwibm8tZW1wdHkiOlsiZXJyb3IiXSwibm8tZW1wdHktY2hhcmFjdGVyLWNsYXNzIjpbImVycm9yIl0sIm5vLWVtcHR5LXBhdHRlcm4iOlsiZXJyb3IiXSwibm8tZXgtYXNzaWduIjpbImVycm9yIl0sIm5vLWV4dHJhLWJvb2xlYW4tY2FzdCI6WyJlcnJvciJdLCJuby1leHRyYS1zZW1pIjpbImVycm9yIl0sIm5vLWZhbGx0aHJvdWdoIjpbImVycm9yIl0sIm5vLWZ1bmMtYXNzaWduIjpbImVycm9yIl0sIm5vLWdsb2JhbC1hc3NpZ24iOlsiZXJyb3IiXSwibm8taW1wb3J0LWFzc2lnbiI6WyJlcnJvciJdLCJuby1pbm5lci1kZWNsYXJhdGlvbnMiOlsiZXJyb3IiXSwibm8taW52YWxpZC1yZWdleHAiOlsiZXJyb3IiXSwibm8taXJyZWd1bGFyLXdoaXRlc3BhY2UiOlsiZXJyb3IiXSwibm8tbG9zcy1vZi1wcmVjaXNpb24iOlsiZXJyb3IiXSwibm8tbWlzbGVhZGluZy1jaGFyYWN0ZXItY2xhc3MiOlsiZXJyb3IiXSwibm8tbWl4ZWQtc3BhY2VzLWFuZC10YWJzIjpbImVycm9yIl0sIm5vLW5ldy1zeW1ib2wiOlsiZXJyb3IiXSwibm8tbm9ub2N0YWwtZGVjaW1hbC1lc2NhcGUiOlsiZXJyb3IiXSwibm8tb2JqLWNhbGxzIjpbImVycm9yIl0sIm5vLW9jdGFsIjpbImVycm9yIl0sIm5vLXByb3RvdHlwZS1idWlsdGlucyI6WyJlcnJvciJdLCJuby1yZWRlY2xhcmUiOlsiZXJyb3IiXSwibm8tcmVnZXgtc3BhY2VzIjpbImVycm9yIl0sIm5vLXNlbGYtYXNzaWduIjpbImVycm9yIl0sIm5vLXNldHRlci1yZXR1cm4iOlsiZXJyb3IiXSwibm8tc2hhZG93LXJlc3RyaWN0ZWQtbmFtZXMiOlsiZXJyb3IiXSwibm8tc3BhcnNlLWFycmF5cyI6WyJlcnJvciJdLCJuby10aGlzLWJlZm9yZS1zdXBlciI6WyJlcnJvciJdLCJuby11bmRlZiI6WyJlcnJvciJdLCJuby11bmV4cGVjdGVkLW11bHRpbGluZSI6WyJlcnJvciJdLCJuby11bnJlYWNoYWJsZSI6WyJlcnJvciJdLCJuby11bnNhZmUtZmluYWxseSI6WyJlcnJvciJdLCJuby11bnNhZmUtbmVnYXRpb24iOlsiZXJyb3IiXSwibm8tdW5zYWZlLW9wdGlvbmFsLWNoYWluaW5nIjpbImVycm9yIl0sIm5vLXVudXNlZC1sYWJlbHMiOlsiZXJyb3IiXSwibm8tdXNlbGVzcy1iYWNrcmVmZXJlbmNlIjpbImVycm9yIl0sIm5vLXVzZWxlc3MtY2F0Y2giOlsiZXJyb3IiXSwibm8tdXNlbGVzcy1lc2NhcGUiOlsiZXJyb3IiXSwibm8td2l0aCI6WyJlcnJvciJdLCJyZXF1aXJlLXlpZWxkIjpbImVycm9yIl0sInVzZS1pc25hbiI6WyJlcnJvciJdLCJ2YWxpZC10eXBlb2YiOlsiZXJyb3IiXX0sImVudiI6eyJlczYiOnRydWV9fX0=)

```js
/* eslint no-shadow: ["error", {"ignoreOnInitialization": true}] */

// should fail
const items1 = [1, 2, 3, 4];
const item1 = 1;
const _item1 = items1.filter((item1) => (item1 % 2)); // error

// should pass
const items2 = [1, 2, 3, 4];
const item2 = items2.filter((item2) => (item2 % 2)); // no-error
```

**Comment by ximex** [comment]
sorry only checked the typescript-eslint docs. this is exactly what it searched for. thx
